### ISSUE TITLE
Pub commands fail with "Invalid tar entry: .."

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/pub#4305]
# Problem

Pub commands (`pub get`, `pub outdated`, `pub global activate`, etc.) fail with output like:

```
Downloading packages... 
Failed to extract `/tmp/pub_FTQXCZ/devtools_shared-10.0.0.tar.gz`: Invalid tar entry: ..
```
### Expected behavior

Pub commands complete.

### Actual behavior

Pub commands fail.

# Cause

Tar files downloaded from pub.dev fail to be extracted due to containing a `.` entry.

This is a severe regression, likely caused by https://github.com/dart-lang/pub/pull/4283

# Exposure

The regression is currently in the Dart `main` and `dev` channels. It is *not* in the `beta` and `stable` channels.

Only the following dev-releases are affected:

* 3.5.0-243.0.dev
* 3.5.0-244.0.dev
* 3.5.0-245.0.dev
* 3.5.0-246.0.dev
* 3.5.0-247.0.dev
* 3.5.0-248.0.dev
* 3.5.0-249.0.dev
* 3.5.0-250.0.dev
* 3.5.0-251.0.dev
* 3.5.0-252.0.dev
* 3.5.0-253.0.dev
* 3.5.0-254.0.dev
* 3.5.0-255.0.dev
* 3.5.0-256.0.dev
* 3.5.0-257.0.dev
* 3.5.0-258.0.dev

The pub.dev server was patched on June 13th to reject publishing further packages with broken archives.

The affected already published package-versions are:

* ant_media_flutter-1.4.1.tar.gz
* cxgenie-2.4.5.tar.gz
* cxgenie-2.4.4.tar.gz
* cxgenie-2.4.3.tar.gz
* cxgenie-2.4.2.tar.gz
* macros-0.1.2-main.1.tar.gz
* devtools_extensions-0.2.0.tar.gz
* devtools_shared-10.0.0.tar.gz
* devtools_app_shared-0.2.0.tar.gz

# Mitigation

Package versions already published to pub.dev will not work. We recommend that package owners use the pub.dev Admin functionality to *retract* these package versions: https://dart.dev/tools/pub/publishing#retract

New package versions can be published safely as long as you use a beta or stable SDK to do the publishing with.

**Comment by brianquinlan** [comment]
I'm not sure how the macro system works. Could. reverting https://github.com/dart-lang/sdk/commit/fac11694b4453808110c8b070de6a4b0182a7318 fix this?

**Comment by parlough** [comment]
I'm getting this with `devtools_shared` as well:

```
Failed to extract `/private/var/folders/[removed]/devtools_shared-10.0.0.tar.gz`: Invalid tar entry: ..
```

**Comment by dcharkes** [comment]
This seems to be unrelated to the macros package, I'm seeing the same as @parlough.

```
Failed to extract `/b/s/w/ir/x/t/pub_DWJXDV/devtools_shared-10.0.0.tar.gz`: Invalid tar entry: ..
```

**Comment by sigurdm** [comment]
Must be a problem with https://github.com/dart-lang/pub/pull/4283

**Comment by sigurdm** [comment]
We are also not setting a mode on the directories. makes the tarballs annoying if you extract with system `tar` (in the client we ignore the permissions from the tar-file except for the x-bit).

**Comment by simolus3** [comment]
Looking at the extraction logic, the problem does not appear to be the mode bits on directories (which are odd, but also completely ignored by `extractTarGz`) which merely calls `ensureDir`.
The problem is that we want to reject archives with entries pointing out of their target directory (e.g. if I upload a malicious tarball with an entry named `../some_other_package-1.2.3/lib/injection.dart` to pub.dev, the client must reject it instead of potentially overwriting the local copy of another package). To do this, we currently abort if `!path.isWithin(destination, tarEntry.name)`. That's correct in almost all cases, but breaks with a top-level directory entry simply named `.`, as that leads to `fileName == destination` and `isWithin` returning false. I think this is a bug in the extraction logic, and I suggest the following patch which fixes the problem for `devtools_shared-10.0.0.tar.gz`, the only package I've tested:

```patch
diff --git a/lib/src/io.dart b/lib/src/io.dart
index 730962b5..0d764eb5 100644
--- a/lib/src/io.dart
+++ b/lib/src/io.dart
@@ -1072,7 +1072,8 @@ Future<void> extractTarGz(Stream<List<int>> stream, String destination) async {
       throw FormatException('Tar file contained duplicate path ${entry.name}');
     }
 
-    if (!p.isWithin(destination, filePath)) {
+    if (!(p.isWithin(destination, filePath) ||
+        (p.equals(destination, filePath) && entry.type == TypeFlag.dir))) {
       // The tar contains entries that would be written outside of the
       // destination. That doesn't happen by accident, assume that the tar file
       // is malicious.

```

Considering that it breaks compatibility with older pub clients, perhaps adding the top-level `.` directory should be skipped. But fixing `extractTarGz` certainly doesn't hurt.

**Comment by sigurdm** [comment]
All affected package versions have been removed from pub.dev
